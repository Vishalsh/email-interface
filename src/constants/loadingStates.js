const loadingStates = {
  AT_REST: 'AT_REST',
  LOADING: 'LOADING',
  LOADED: 'LOADED',
  FAILED: 'FAILED'
};

export default loadingStates;