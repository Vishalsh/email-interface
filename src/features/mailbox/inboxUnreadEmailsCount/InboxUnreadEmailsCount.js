import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Badge from 'components/badge/Badge';
import inboxUnreadEmailsCountSelector from './inboxUnreadEmailsCountSelector';
import classes from './InboxUnreadEmailsCount.module.scss';

export const InboxUnreadEmailsCount = (props) => {
  const {
    readEmailsCount,
    unreadEmailsCount,
    showReadMailsCount,
    showBadge
  } = props;

  const label = showReadMailsCount ? `${unreadEmailsCount}/${readEmailsCount}` : unreadEmailsCount;

  return (
    unreadEmailsCount !== 0 &&
    <>
    {
      showBadge ?
        <Badge className={classes.inboxUnreadEmailsCount} label={label}/>
        :
        `(${label})`
    }
    </>
  );
};

InboxUnreadEmailsCount.defaultProps = {
  showReadMailsCount: false,
  showBadge: true
};

InboxUnreadEmailsCount.propTypes = {
  readEmailsCount: PropTypes.number.isRequired,
  unreadEmailsCount: PropTypes.number.isRequired,
  showReadMailsCount: PropTypes.bool,
  showBadge: PropTypes.bool
};

const mapStateToProps = (store) => ({
  readEmailsCount: store.mailbox.inbox ? store.mailbox.inbox.emails.length : 0,
  unreadEmailsCount: inboxUnreadEmailsCountSelector(store),
});

export default connect(mapStateToProps)(InboxUnreadEmailsCount);
