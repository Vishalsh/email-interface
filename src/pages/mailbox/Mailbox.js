import React from 'react';
import { Route, Switch, Redirect } from 'react-router-dom';

import routes from 'constants/routes';
import SettingsCog from 'components/settingsCog/SettingsCog';
import MailboxAside from 'features/mailboxAside/MailboxAside';
import MailboxFeature from 'features/mailbox/Mailbox';

const Mailbox = (props) => {
  const { match } = props;

  return (
    <div className="row">
      <div className="col-sm-3">
        <MailboxAside />
      </div>
      <div className="col-sm-9">
        <Switch>
          <Redirect exact from={routes.MAILBOX} to={routes.INBOX}/>
          <Route path={`${match.url}/:mailbox`} component={MailboxFeature}/>
        </Switch>
      </div>
      <SettingsCog />
    </div>
  );
};

export default Mailbox;
