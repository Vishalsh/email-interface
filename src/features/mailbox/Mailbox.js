import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux'
import { Route } from 'react-router-dom';

import mailboxActions from './mailbox.actions';
import loadingStates from 'constants/loadingStates';
import Emails from 'features/emails/Emails';
import NoEmail from 'features/emails/noEmail/NoEmail';
import EmailDetails from 'features/emails/emailDetails/EmailDetails';
import CreateEmail from 'features/emails/createEmail/CreateEmail';
import MailboxHeader from 'features/mailbox/mailboxHeader/MailboxHeader';
import classes from './Mailbox.module.scss';

export class Mailbox extends Component {
  componentDidMount() {
    const { match, getEmails, mailbox } = this.props;

    if (mailbox.loadingState === loadingStates.AT_REST) {
      getEmails(match.params.mailbox);
    }
  }

  componentWillReceiveProps(nextProps) {
    const { match, getEmails, mailbox } = nextProps;

    if ((match.params.mailbox !== this.props.match.params.mailbox) && mailbox.loadingState === loadingStates.AT_REST) {
      getEmails(match.params.mailbox);
    }
  }

  onClickDeleteEmails = () => {
    const { match, deleteEmails } = this.props;

    deleteEmails(match.params.mailbox);
  };

  render() {
    const { match, mailbox, isCreateEmailPopupOpen } = this.props;

    return (
      <section className={classes.mailbox}>
        <MailboxHeader mailbox={match.params.mailbox} onClickDeleteEmails={this.onClickDeleteEmails}/>

        {
          mailbox.emails.length > 0 ?
            <Emails mailbox={match.params.mailbox} ids={mailbox.emails}/>
            :
            <NoEmail />
        }

        <Route path={`${match.url}/:id`} component={EmailDetails}/>

        {
          isCreateEmailPopupOpen &&
          <CreateEmail />
        }
      </section>
    );
  }
}

Mailbox.defaultProps = {
  mailbox: {
    emails: [],
    loadingState: loadingStates.AT_REST
  },
  isCreateEmailPopupOpen: false
};

Mailbox.propTypes = {
  match: PropTypes.object.isRequired,
  mailbox: PropTypes.shape({
    emails: PropTypes.array,
    loadingState: PropTypes.string,
  }),
  isCreateEmailPopupOpen: PropTypes.bool.isRequired,
  getEmails: PropTypes.func.isRequired,
  deleteEmails: PropTypes.func.isRequired,
};

const mapStateToProps = (store, ownProps) => ({
  mailbox: store.mailbox[ownProps.match.params.mailbox],
  isCreateEmailPopupOpen: store.emails.isCreateEmailPopupOpen
});

const mapDispatchToProps = dispatch => bindActionCreators({
  getEmails: mailboxActions.getEmails,
  deleteEmails: mailboxActions.deleteEmails
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Mailbox);
