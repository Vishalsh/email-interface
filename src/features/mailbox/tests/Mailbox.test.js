import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import loadingStates from 'constants/loadingStates';

import InboxUnreadEmailsCount from 'features/mailbox/inboxUnreadEmailsCount/InboxUnreadEmailsCount';
import { Mailbox } from '../Mailbox';

describe('Mailbox', () => {
  const getEmailsMock = jest.fn();
  const props = {
    match: {
      params: {
        mailbox: 'inbox'
      },
      url: '/mailbox/inbox'
    },
    mailbox: {
      emails: [1, 2],
      loadingState: loadingStates.LOADED
    },
    getEmails: getEmailsMock
  };
  const component = shallow(<Mailbox {...props} />);

  it('should render the component', () => {
    expect(toJson(component)).toMatchSnapshot();
    expect(getEmailsMock).not.toHaveBeenCalled();
  });

  it('should trigger getEmails for the mailbox when not present', () => {
    const propsWithoutEmails = {
      ...props,
      mailbox: undefined
    };
    shallow(<Mailbox {...propsWithoutEmails} />);

    expect(getEmailsMock).toHaveBeenCalled();
  });

  it('should unread emails count only for Inbox', () => {
    expect(component.find(InboxUnreadEmailsCount)).toHaveLength(1);

    const propWithSentMailbox = {
      ...props,
      match: {
        params: {
          mailbox: 'sent'
        },
        url: '/mailbox/sent'
      }
    };
    const componentWithSentMailbox = shallow(<Mailbox {...propWithSentMailbox} />);

    expect(componentWithSentMailbox.find(InboxUnreadEmailsCount)).toHaveLength(0);
  });

  it('should trigger getEmails on change of mailbox when already not present', () => {
    getEmailsMock.mockReset();

    component.setProps({
      match: {
        params: {
          mailbox: 'inbox'
        }
      },
      mailbox: {
        emails: [1, 2],
        loadingState: loadingStates.LOADED
      },
    });

    expect(getEmailsMock).not.toHaveBeenCalled();

    component.setProps({
      match: {
        params: {
          mailbox: 'sent'
        }
      },
      mailbox: {
        emails: [1, 2],
        loadingState: loadingStates.LOADED
      },
    });

    expect(getEmailsMock).not.toHaveBeenCalled();

    component.setProps({
      match: {
        params: {
          mailbox: 'trash'
        }
      },
      mailbox: {
        emails: [],
        loadingState: loadingStates.AT_REST
      }
    });

    expect(getEmailsMock).toHaveBeenCalled();
  });
});