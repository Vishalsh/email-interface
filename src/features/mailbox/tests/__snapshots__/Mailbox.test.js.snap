// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mailbox should render the component 1`] = `
<section
  className="mailbox"
>
  <header
    className="header"
  >
    <div
      className="row"
    >
      <h1
        className="name col-sm-6"
      >
        inbox
        <Connect(InboxUnreadEmailsCount)
          showBadge={false}
        />
      </h1>
    </div>
    <div
      className="mailboxActions row"
    >
      <div
        className="col-sm-8 flex"
      >
        <Button
          className="actionButton"
          disabled={false}
          type="default"
        >
          <i
            className="icon icon-refresh iconRefresh"
          />
          <span>
            Refresh
          </span>
        </Button>
        <Button
          className="actionButton"
          disabled={false}
          type="default"
        >
          <i
            className="icon icon-eye"
          />
        </Button>
        <Button
          className="actionButton"
          disabled={false}
          type="default"
        >
          <i
            className="icon icon-bang"
          />
        </Button>
        <Button
          className="actionButton"
          disabled={false}
          onClick={[Function]}
          type="default"
        >
          <i
            className="icon icon-delete"
          />
        </Button>
      </div>
      <div
        className="col-sm-4 end-xs flex"
      >
        <Button
          className="actionButton"
          disabled={false}
          type="default"
        >
          <i
            className="icon icon-left"
          />
        </Button>
        <Button
          className="actionButton paginationRightButton"
          disabled={false}
          type="default"
        >
          <i
            className="icon icon-right"
          />
        </Button>
      </div>
    </div>
  </header>
  <Connect(Emails)
    ids={
      Array [
        1,
        2,
      ]
    }
    mailbox="inbox"
  />
  <Route
    component={[Function]}
    path="/mailbox/inbox/:id"
  />
</section>
`;
